library/core/src/ptr/non_null.rs

```diff
@@ -1829,4 +1856,103 @@ mod verify {
+        // unit type
+        let ptr_unit = NonNull::<()>::dangling();
+        // zero length slice from dangling unit pointer
+        let zero_len_slice =  NonNull::slice_from_raw_parts(ptr_unit, 0);
+    }
+
+    // pub const fn from_raw_parts(data_pointer: NonNull<()>, metadata: <T as super::Pointee>::Metadata,) -> NonNull<T>
+    #[kani::proof_for_contract(NonNull::from_raw_parts)]
+    pub fn non_null_check_from_raw_parts() {
+
+        const ARR_LEN: usize = 10000;
+        // Create a non-deterministic array and its slice
+        let arr: [i8; ARR_LEN] = kani::any();
+        let arr_slice = &arr[..];  
+        // Get a raw NonNull pointer to the start of the slice
+        let arr_slice_raw_ptr = NonNull::new(arr_slice.as_ptr() as *mut ()).unwrap();  
+        // Create NonNull pointer from the start pointer and the length of the slice
+        let nonnull_slice = NonNull::<[i8]>::from_raw_parts(arr_slice_raw_ptr, ARR_LEN);
+        // Ensure slice content is preserved, runtime at this step is proportional to ARR_LEN
+        unsafe {
+            kani::assert( arr_slice == nonnull_slice.as_ref(), "slice content must preserve" );
+        }
+            
+        // Create a SampleTrait object from SampleStruct
+        let sample_struct = SampleStruct { value: kani::any() };
+        let trait_object: &dyn SampleTrait = &sample_struct;
+
+        // Get the raw data pointer and metadata for the trait object
+        let trait_ptr = NonNull::new(trait_object as *const dyn SampleTrait as *mut ()).unwrap();
+        // Safety: For trait objects, the metadata must come from a pointer to the same underlying erased type
+        let metadata = core::ptr::metadata(trait_object);
+
+        // Create NonNull<dyn MyTrait> from the data pointer and metadata
+        let nonnull_trait_object: NonNull<dyn SampleTrait> = NonNull::from_raw_parts(trait_ptr, metadata);
+
+        unsafe {
+            // Ensure trait method and member is preserved
+            kani::assert( trait_object.get_value() == nonnull_trait_object.as_ref().get_value(), "trait method and member must correctly preserve");
+        }
+
+    }
+
+    // pub const fn slice_from_raw_parts(data: NonNull<T>, len: usize) -> Self
+    #[kani::proof_for_contract(NonNull::slice_from_raw_parts)]
+    #[kani::unwind(1001)]
+    pub fn non_null_check_slice_from_raw_parts() {
+        const ARR_LEN: usize = 1000;
+        // Create a non-deterministic array
+        let mut arr: [i8; ARR_LEN] = kani::any();
+        // Get a raw NonNull pointer to the start of the slice
+        let arr_raw_ptr = NonNull::new(arr.as_mut_ptr()).unwrap();  
+        // Create NonNull slice from the start pointer and ends at random slice_len
+        // Safety: https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html
+        let slice_len: usize = kani::any();
+        kani::assume(slice_len <= ARR_LEN);
+        let nonnull_slice = NonNull::<[i8]>::slice_from_raw_parts(arr_raw_ptr, slice_len);
```

Comment:
Same as the previous harness.

Reply:
@qinheping If I add the zero-length pointer in the harness like this:
```rust
        // zero-length dangling pointer example
        let dangling_ptr = NonNull::<()>::dangling();
        let zero_length = NonNull::<[()]>::slice_from_raw_parts(dangling_ptr, 0);
```
It results the following failure:
```
Check 72: kani::mem::cbmc::same_allocation::<()>.assertion.1
	 - Status: FAILURE
	 - Description: "Kani does not support reasoning about pointer to unallocated memory"
	 - Location: library/core/src/lib.rs:423:1 in function kani::mem::cbmc::same_allocation::<()>
```
Due to this contract clause:
```rust
kani::mem::same_allocation(data.pointer, data.pointer.add(len)
```

Reply:
Also, since we have:
```rust
        let slice_len: usize = kani::any();
        kani::assume(slice_len <= ARR_LEN);
```
Does this mean that we are already including the case where `slice_len == 0`?

Reply:
Sorry I forgot that the issue of dangling pointers is not fixed yet: https://github.com/model-checking/kani/issues/3670.
Please commented out the dangling pointer test and leave a comment as TODO so that we can improve the harness later.

Then I think using `any_slice_from_array` would be enough for now. Could you figure out why it fails the proof? Do we miss any preconditions in the contracts?

Reply:
Commented out this example